<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#DATAMART" data-toc-modified-id="DATAMART-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>DATAMART</a></span><ul class="toc-item"><li><span><a href="#Set-up" data-toc-modified-id="Set-up-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Set up</a></span></li><li><span><a href="#Loading-data" data-toc-modified-id="Loading-data-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Loading data</a></span></li><li><span><a href="#Data-integration" data-toc-modified-id="Data-integration-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Data integration</a></span></li><li><span><a href="#Data-quality" data-toc-modified-id="Data-quality-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Data quality</a></span><ul class="toc-item"><li><span><a href="#Types-of-variables" data-toc-modified-id="Types-of-variables-1.4.1"><span class="toc-item-num">1.4.1&nbsp;&nbsp;</span>Types of variables</a></span></li><li><span><a href="#Null-analysis" data-toc-modified-id="Null-analysis-1.4.2"><span class="toc-item-num">1.4.2&nbsp;&nbsp;</span>Null analysis</a></span></li><li><span><a href="#Analysis-of-numeric-variables" data-toc-modified-id="Analysis-of-numeric-variables-1.4.3"><span class="toc-item-num">1.4.3&nbsp;&nbsp;</span>Analysis of numeric variables</a></span></li><li><span><a href="#Analysis-of-categorical-variables" data-toc-modified-id="Analysis-of-categorical-variables-1.4.4"><span class="toc-item-num">1.4.4&nbsp;&nbsp;</span>Analysis of categorical variables</a></span></li><li><span><a href="#Index" data-toc-modified-id="Index-1.4.5"><span class="toc-item-num">1.4.5&nbsp;&nbsp;</span>Index</a></span></li></ul></li><li><span><a href="#Data-transformation" data-toc-modified-id="Data-transformation-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Data transformation</a></span><ul class="toc-item"><li><span><a href="#Components-of-the-date" data-toc-modified-id="Components-of-the-date-1.5.1"><span class="toc-item-num">1.5.1&nbsp;&nbsp;</span>Components of the date</a></span></li><li><span><a href="#Calendar-vairables:-holidays" data-toc-modified-id="Calendar-vairables:-holidays-1.5.2"><span class="toc-item-num">1.5.2&nbsp;&nbsp;</span>Calendar vairables: holidays</a></span></li><li><span><a href="#Exogenous-indicators" data-toc-modified-id="Exogenous-indicators-1.5.3"><span class="toc-item-num">1.5.3&nbsp;&nbsp;</span>Exogenous indicators</a></span></li></ul></li><li><span><a href="#FINAL-ANALYTICAL-BOARD" data-toc-modified-id="FINAL-ANALYTICAL-BOARD-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>FINAL ANALYTICAL BOARD</a></span></li></ul></li></ul></div>

# DATAMART 

## Set up

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

%config IPCompleter.greedy = True
pd.options.display.float_format = '{:15.2f}'.format


## Loading data

In [3]:
import sqlalchemy as sa
con = sa.create_engine('sqlite:///../Datos/ecommerce.db')

In [4]:
from sqlalchemy import inspect
insp = inspect(con)
insp.get_table_names()

['2019-Dec', '2019-Nov', '2019-Oct', '2020-Feb', '2020-Jan']

In [5]:
octo = pd.read_sql('2019-Oct', con)
nov = pd.read_sql('2019-Nov', con)
dec = pd.read_sql('2019-Dec', con)
jan = pd.read_sql('2020-Jan' ,con)
feb = pd.read_sql('2020-Feb' ,con)

## Data integration

In [6]:
octo.head()

index               event_time event_type  product_id          category_id  \
0     68  2019-10-01 00:01:46 UTC       view     5843665  1487580005092295511   
1     72  2019-10-01 00:01:55 UTC       cart     5868461  1487580013069861041   
2     95  2019-10-01 00:02:50 UTC       view     5877456  1487580006300255120   
3    122  2019-10-01 00:03:41 UTC       view     5649270  1487580013749338323   
4    124  2019-10-01 00:03:44 UTC       view       18082  1487580005411062629   

  category_code     brand           price    user_id  \
0          None     f.o.x            9.44  462033176   
1          None   italwax            3.57  514753614   
2          None  jessnail          122.22  527418424   
3          None   concept            6.19  555448072   
4          None       cnd           16.03  552006247   

                           user_session  
0  a18e0999-61a1-4218-8f8f-61ec1d375361  
1  e2fecb2d-22d0-df2c-c661-15da44b3ccf1  
2  86e77869-afbc-4dff-9aa2-6b7dd8c90770  
3  b5f72ceb-0730-44de-a932-d16db62390df  
4  2d8f304b-de45-4e59-8f40-50c603843fe5

In [7]:
nov.head()

index               event_time event_type  product_id          category_id  \
0     75  2019-11-01 00:04:51 UTC       view     5861591  1487580009143992338   
1     97  2019-11-01 00:06:55 UTC       view     5747404  1487580005671109489   
2    101  2019-11-01 00:07:31 UTC       view     5747406  1487580005671109489   
3    126  2019-11-01 00:10:03 UTC       view        4607  1487580010796548183   
4    139  2019-11-01 00:11:12 UTC       view     5866135  1487580008263189483   

  category_code   brand           price    user_id  \
0          None   lador            2.22  534488348   
1          None  masura            6.33  554428484   
2          None  masura            6.33  554428484   
3          None  runail            0.38  566281978   
4          None  s.care          103.17  566282124   

                           user_session  
0  25360ead-b337-47b1-a61d-e684a8491179  
1  64ccbe91-91f6-4671-b276-857dfa0e99aa  
2  64ccbe91-91f6-4671-b276-857dfa0e99aa  
3  dae10616-018b-404b-ab5e-76166fb17bbe  
4  51d54818-54fc-49fd-bb4b-9553ea4bdc97

In [8]:
df = pd.concat([octo,nov,dec,jan,feb], axis = 0)
df

index               event_time event_type  product_id  \
0            68  2019-10-01 00:01:46 UTC       view     5843665   
1            72  2019-10-01 00:01:55 UTC       cart     5868461   
2            95  2019-10-01 00:02:50 UTC       view     5877456   
3           122  2019-10-01 00:03:41 UTC       view     5649270   
4           124  2019-10-01 00:03:44 UTC       view       18082   
...         ...                      ...        ...         ...   
429785  4156660  2020-02-29 23:58:49 UTC       cart     5815662   
429786  4156663  2020-02-29 23:58:57 UTC       view     5815665   
429787  4156668  2020-02-29 23:59:05 UTC       cart     5815665   
429788  4156675  2020-02-29 23:59:28 UTC       view     5817692   
429789  4156680  2020-02-29 23:59:54 UTC       view     5716351   

                category_id category_code     brand           price  \
0       1487580005092295511          None     f.o.x            9.44   
1       1487580013069861041          None   italwax            3.57   
2       1487580006300255120          None  jessnail          122.22   
3       1487580013749338323          None   concept            6.19   
4       1487580005411062629          None       cnd           16.03   
...                     ...           ...       ...             ...   
429785  1487580006317032337          None      None            0.92   
429786  1487580006317032337          None      None            0.59   
429787  1487580006317032337          None      None            0.59   
429788  1487580010872045658          None      None            0.79   
429789  1487580010872045658          None     irisk            0.79   

          user_id                          user_session  
0       462033176  a18e0999-61a1-4218-8f8f-61ec1d375361  
1       514753614  e2fecb2d-22d0-df2c-c661-15da44b3ccf1  
2       527418424  86e77869-afbc-4dff-9aa2-6b7dd8c90770  
3       555448072  b5f72ceb-0730-44de-a932-d16db62390df  
4       552006247  2d8f304b-de45-4e59-8f40-50c603843fe5  
...           ...                                   ...  
429785  147995998  5ff96629-3627-493e-a25b-5a871ec78c90  
429786  147995998  5ff96629-3627-493e-a25b-5a871ec78c90  
429787  147995998  5ff96629-3627-493e-a25b-5a871ec78c90  
429788  619841242  18af673b-7fb9-4202-a66d-5c855bc0fd2d  
429789  619841242  18af673b-7fb9-4202-a66d-5c855bc0fd2d  

[2095076 rows x 10 columns]

## Data quality

### Types of variables

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2095076 entries, 0 to 429789
Data columns (total 10 columns):
 #   Column         Dtype  
---  ------         -----  
 0   index          int64  
 1   event_time     object 
 2   event_type     object 
 3   product_id     int64  
 4   category_id    int64  
 5   category_code  object 
 6   brand          object 
 7   price          float64
 8   user_id        int64  
 9   user_session   object 
dtypes: float64(1), int64(4), object(5)
memory usage: 175.8+ MB


We delete the column index because for this analysis is not necessary 

In [10]:
df.drop(columns = 'index', inplace = True)

I don't know why when you use a large amount of data the function pd.to_datatime the time to do the conversion is too long (approx 5 minutes). To solve this minor problem, we can use the next function that makes the process in less than 15 sec.

In [11]:
#df['event_time'] = pd.to_datetime(df['event_time'])
#This method consum a lot of time (aprox 5 minutes) compared with the function datetime_quickly

In [12]:
def datetime_quickly(df, formato):
    '''
    This function help us to convert very quickly a big numbers of registers in a datetime format
    
    '''
    def split_date(date):
        division = date.split()
        date = division[0]
        time = division[1]
        cadena = date + ' ' + time
        return (cadena)
    result = pd.to_datetime(df.apply( lambda x: split_date(x)), format = formato)
    return(result)
        

In [13]:
formato = '%Y-%m-%d %H:%M:%S'
df.event_time = datetime_quickly(df['event_time'], formato)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2095076 entries, 0 to 429789
Data columns (total 9 columns):
 #   Column         Dtype         
---  ------         -----         
 0   event_time     datetime64[ns]
 1   event_type     object        
 2   product_id     int64         
 3   category_id    int64         
 4   category_code  object        
 5   brand          object        
 6   price          float64       
 7   user_id        int64         
 8   user_session   object        
dtypes: datetime64[ns](1), float64(1), int64(3), object(4)
memory usage: 159.8+ MB


### Null analysis

In [14]:
df.isna().sum().sort_values(ascending = False)

category_code    2060411
brand             891646
user_session         506
event_time             0
event_type             0
product_id             0
category_id            0
price                  0
user_id                0
dtype: int64

Conclusions:

* category_code has almost all records to null
* brand has almost half of the records set to null
* there are 500 nulls in session

Actions:

* delete the variables categoria_cod and brand
* remove session nulls as it is a relevant variable

In [15]:
df= df.drop(columns = ['category_code','brand']).dropna()

In [16]:
df

event_time event_type  product_id          category_id  \
0      2019-10-01 00:01:46       view     5843665  1487580005092295511   
1      2019-10-01 00:01:55       cart     5868461  1487580013069861041   
2      2019-10-01 00:02:50       view     5877456  1487580006300255120   
3      2019-10-01 00:03:41       view     5649270  1487580013749338323   
4      2019-10-01 00:03:44       view       18082  1487580005411062629   
...                    ...        ...         ...                  ...   
429785 2020-02-29 23:58:49       cart     5815662  1487580006317032337   
429786 2020-02-29 23:58:57       view     5815665  1487580006317032337   
429787 2020-02-29 23:59:05       cart     5815665  1487580006317032337   
429788 2020-02-29 23:59:28       view     5817692  1487580010872045658   
429789 2020-02-29 23:59:54       view     5716351  1487580010872045658   

                 price    user_id                          user_session  
0                 9.44  462033176  a18e0999-61a1-4218-8f8f-61ec1d375361  
1                 3.57  514753614  e2fecb2d-22d0-df2c-c661-15da44b3ccf1  
2               122.22  527418424  86e77869-afbc-4dff-9aa2-6b7dd8c90770  
3                 6.19  555448072  b5f72ceb-0730-44de-a932-d16db62390df  
4                16.03  552006247  2d8f304b-de45-4e59-8f40-50c603843fe5  
...                ...        ...                                   ...  
429785            0.92  147995998  5ff96629-3627-493e-a25b-5a871ec78c90  
429786            0.59  147995998  5ff96629-3627-493e-a25b-5a871ec78c90  
429787            0.59  147995998  5ff96629-3627-493e-a25b-5a871ec78c90  
429788            0.79  619841242  18af673b-7fb9-4202-a66d-5c855bc0fd2d  
429789            0.79  619841242  18af673b-7fb9-4202-a66d-5c855bc0fd2d  

[2094570 rows x 7 columns]

### Analysis of numeric variables

In [17]:
df.describe().T

count                   mean                   std  \
product_id       2094570.00             5487103.56            1300923.90   
category_id      2094570.00 1553112489402986240.00 167907497920779328.00   
price            2094570.00                   8.42                 19.14   
user_id          2094570.00           521077545.56           87553855.76   

                               min                    25%  \
product_id                 3752.00             5724652.00   
category_id 1487580004807082752.00 1487580005754995456.00   
price                       -47.62                   2.05   
user_id                 4661182.00           480613387.00   

                               50%                    75%  \
product_id              5811665.00             5858353.00   
category_id 1487580008246412288.00 1487580013489291520.00   
price                         4.00                   6.86   
user_id               553341613.00           578406571.00   

                               max  
product_id              5932595.00  
category_id 2242903426784559104.00  
price                       327.78  
user_id               622087993.00

We see a strange value in the price, the min value is a negative price.

In the context of retail products, you can find negative values that are related to returns of products by the client. But this value is often in a different column than the price.

We look deeper into the data

In [18]:
df[df.price <= 0]

event_time event_type  product_id          category_id  \
343    2019-10-01 02:15:41       view     5892052  1487580010377117763   
924    2019-10-01 05:16:30       view     5889621  1487580010561667147   
933    2019-10-01 05:18:03       view     5889622  1487580010561667147   
937    2019-10-01 05:18:46       view     5889623  1487580010561667147   
1077   2019-10-01 05:38:01       view     5889627  1487580010561667147   
...                    ...        ...         ...                  ...   
428011 2020-02-29 20:04:49       cart     5824841  1897124478404526487   
428012 2020-02-29 20:04:49       cart     5826413  1487580005511725929   
428013 2020-02-29 20:04:49       cart     5832437  1487580007675986893   
428014 2020-02-29 20:04:49       cart     5851606  2055161088059638328   
428370 2020-02-29 20:26:16       view     5923106  1487580008246412266   

                 price    user_id                          user_session  
343               0.00  555455025  320f6021-30ac-4a58-ae17-bac1cc32aac3  
924               0.00  523988665  00849bd2-fcd2-4cb4-af31-4e264f151848  
933               0.00  523988665  80cfe614-f0a5-4101-a2b6-a21227590470  
937               0.00  523988665  c2cd0464-3d2b-48e2-9667-bac248fe297a  
1077              0.00  523988665  8b2bf9d8-43f0-43b2-bed3-13b2c956cada  
...                ...        ...                                   ...  
428011            0.00  469761446  8bf369b4-92c0-4fb8-88a5-8a2dd0947e46  
428012            0.00  469761446  8bf369b4-92c0-4fb8-88a5-8a2dd0947e46  
428013            0.00  469761446  8bf369b4-92c0-4fb8-88a5-8a2dd0947e46  
428014            0.00  469761446  8bf369b4-92c0-4fb8-88a5-8a2dd0947e46  
428370            0.00  622047714  74f04dc6-2b3c-4565-beda-f575d73ed81c  

[20544 rows x 7 columns]

In [19]:
df[df.price <= 0].product_id.value_counts()


5896186    79
5903915    50
5873428    37
5851294    29
5851304    29
           ..
5859332     1
5859329     1
5858607     1
5848998     1
5832437     1
Name: product_id, Length: 7049, dtype: int64

We see that the error is in different products. The problem is not in a singular product.

In [20]:
df = df[df.price > 0]
df

event_time event_type  product_id          category_id  \
0      2019-10-01 00:01:46       view     5843665  1487580005092295511   
1      2019-10-01 00:01:55       cart     5868461  1487580013069861041   
2      2019-10-01 00:02:50       view     5877456  1487580006300255120   
3      2019-10-01 00:03:41       view     5649270  1487580013749338323   
4      2019-10-01 00:03:44       view       18082  1487580005411062629   
...                    ...        ...         ...                  ...   
429785 2020-02-29 23:58:49       cart     5815662  1487580006317032337   
429786 2020-02-29 23:58:57       view     5815665  1487580006317032337   
429787 2020-02-29 23:59:05       cart     5815665  1487580006317032337   
429788 2020-02-29 23:59:28       view     5817692  1487580010872045658   
429789 2020-02-29 23:59:54       view     5716351  1487580010872045658   

                 price    user_id                          user_session  
0                 9.44  462033176  a18e0999-61a1-4218-8f8f-61ec1d375361  
1                 3.57  514753614  e2fecb2d-22d0-df2c-c661-15da44b3ccf1  
2               122.22  527418424  86e77869-afbc-4dff-9aa2-6b7dd8c90770  
3                 6.19  555448072  b5f72ceb-0730-44de-a932-d16db62390df  
4                16.03  552006247  2d8f304b-de45-4e59-8f40-50c603843fe5  
...                ...        ...                                   ...  
429785            0.92  147995998  5ff96629-3627-493e-a25b-5a871ec78c90  
429786            0.59  147995998  5ff96629-3627-493e-a25b-5a871ec78c90  
429787            0.59  147995998  5ff96629-3627-493e-a25b-5a871ec78c90  
429788            0.79  619841242  18af673b-7fb9-4202-a66d-5c855bc0fd2d  
429789            0.79  619841242  18af673b-7fb9-4202-a66d-5c855bc0fd2d  

[2074026 rows x 7 columns]

Considering that this is a Discovery analysis we decided to eliminate the 0 and negative prices. In a real case you could talk to the ecommerce to try to understand why there are negative values.

### Analysis of categorical variables

In [21]:
df.event_type.nunique()

4

In [22]:
df.event_type.value_counts()

view                961558
cart                574547
remove_from_cart    410357
purchase            127564
Name: event_type, dtype: int64

In [23]:
df.product_id.nunique()

45327

In [24]:
df.category_id.nunique()

508

We have four types of events in the dataframe.

We have 45327 products

We have 508 categories

### Index

In [25]:
df.set_index('event_time', inplace = True)
df

event_type  product_id          category_id  \
event_time                                                        
2019-10-01 00:01:46       view     5843665  1487580005092295511   
2019-10-01 00:01:55       cart     5868461  1487580013069861041   
2019-10-01 00:02:50       view     5877456  1487580006300255120   
2019-10-01 00:03:41       view     5649270  1487580013749338323   
2019-10-01 00:03:44       view       18082  1487580005411062629   
...                        ...         ...                  ...   
2020-02-29 23:58:49       cart     5815662  1487580006317032337   
2020-02-29 23:58:57       view     5815665  1487580006317032337   
2020-02-29 23:59:05       cart     5815665  1487580006317032337   
2020-02-29 23:59:28       view     5817692  1487580010872045658   
2020-02-29 23:59:54       view     5716351  1487580010872045658   

                              price    user_id  \
event_time                                       
2019-10-01 00:01:46            9.44  462033176   
2019-10-01 00:01:55            3.57  514753614   
2019-10-01 00:02:50          122.22  527418424   
2019-10-01 00:03:41            6.19  555448072   
2019-10-01 00:03:44           16.03  552006247   
...                             ...        ...   
2020-02-29 23:58:49            0.92  147995998   
2020-02-29 23:58:57            0.59  147995998   
2020-02-29 23:59:05            0.59  147995998   
2020-02-29 23:59:28            0.79  619841242   
2020-02-29 23:59:54            0.79  619841242   

                                             user_session  
event_time                                                 
2019-10-01 00:01:46  a18e0999-61a1-4218-8f8f-61ec1d375361  
2019-10-01 00:01:55  e2fecb2d-22d0-df2c-c661-15da44b3ccf1  
2019-10-01 00:02:50  86e77869-afbc-4dff-9aa2-6b7dd8c90770  
2019-10-01 00:03:41  b5f72ceb-0730-44de-a932-d16db62390df  
2019-10-01 00:03:44  2d8f304b-de45-4e59-8f40-50c603843fe5  
...                                                   ...  
2020-02-29 23:58:49  5ff96629-3627-493e-a25b-5a871ec78c90  
2020-02-29 23:58:57  5ff96629-3627-493e-a25b-5a871ec78c90  
2020-02-29 23:59:05  5ff96629-3627-493e-a25b-5a871ec78c90  
2020-02-29 23:59:28  18af673b-7fb9-4202-a66d-5c855bc0fd2d  
2020-02-29 23:59:54  18af673b-7fb9-4202-a66d-5c855bc0fd2d  

[2074026 rows x 6 columns]

## Data transformation

We are going to create 3 types of new variables

* Extract components
* Calendar variables: Local holidays (Russia)
* Exogenous indicators: Days not necessarily holidays but with commercial interest: Black Friday, Cyber Monday, Epiphany, Valentine's Day.

### Components of the date

In [26]:
def date_components(dataframe):
    ''' 
    This function split the components of an index date in their parts.
    '''
    date = dataframe.index.date
    year = dataframe.index.year
    month = dataframe.index.month
    day = dataframe.index.day
    hour = dataframe.index.hour
    minute = dataframe.index.minute
    second = dataframe.index.second
    
    
    return(pd.DataFrame({'date':date, 'year':year,'month':month, 'day':day, 'hour':hour, 'minute':minute, 'second':second}))

In [27]:
df = pd.concat([df.reset_index(), date_components(df)], axis = 1)

### Calendar vairables: holidays

In [28]:
import holidays

In [29]:
holidays_RU = holidays.RU(years = 2020)
holidays_RU

{datetime.date(2020, 1, 1): 'Новый год', datetime.date(2020, 1, 2): 'Новый год', datetime.date(2020, 1, 3): 'Новый год', datetime.date(2020, 1, 4): 'Новый год', datetime.date(2020, 1, 5): 'Новый год', datetime.date(2020, 1, 6): 'Новый год', datetime.date(2020, 1, 7): 'Православное Рождество', datetime.date(2020, 1, 8): 'Новый год', datetime.date(2020, 2, 23): 'День защитника отечества', datetime.date(2020, 3, 8): 'День женщин', datetime.date(2020, 5, 1): 'Праздник Весны и Труда', datetime.date(2020, 5, 9): 'День Победы', datetime.date(2020, 6, 12): 'День России', datetime.date(2020, 11, 4): 'День народного единства', datetime.date(2020, 12, 31): 'Новый год'}

In [30]:
df['holidays'] = df.date.apply(lambda x: 1 if (x in holidays_RU) else 0)

In [31]:
df[df.holidays == 1].date.value_counts().sort_index()

2019-11-04    16430
2019-12-31     2848
2020-01-01     7644
2020-01-02    10776
2020-01-03    10617
2020-01-04    13084
2020-01-05    14554
2020-01-06    10621
2020-01-07    12922
2020-01-08    14004
2020-02-23     9817
Name: date, dtype: int64

### Exogenous indicators

We will add indicators for Black Friday and Valentine's Day.

In [ ]:
df.set_index('event_time', inplace = True)

In [37]:
df['black_friday'] = 0
df.loc['2019-11-29', 'black_friday'] = 1

df['valentine_day'] = 0
df.loc['2020-02-14','valentine_day'] = 1


In [46]:
df['black_friday'].value_counts()

0    2051695
1      22331
Name: black_friday, dtype: int64

In [45]:
df['valentine_day'].value_counts()

0    2061781
1      12245
Name: valentine_day, dtype: int64

## FINAL ANALYTICAL BOARD

In [48]:
variables = df.columns.to_list()
variables

['event_type',
 'product_id',
 'category_id',
 'price',
 'user_id',
 'user_session',
 'date',
 'year',
 'month',
 'day',
 'hour',
 'minute',
 'second',
 'holidays',
 'black_friday',
 'valentine_day']

In [49]:
variables_order = ['user_id', 'user_session',  'category_id', 'event_type', 'product_id', 'price' ]
variables_order

['user_id', 'user_session', 'category_id', 'event_type', 'product_id', 'price']

In [52]:
other_variables = [name for name in variables if name not in variables_order]
other_variables

['date',
 'year',
 'month',
 'day',
 'hour',
 'minute',
 'second',
 'holidays',
 'black_friday',
 'valentine_day']

In [55]:
df= df[variables_order + other_variables]
df

user_id                          user_session  \
event_time                                                             
2019-10-01 00:01:46  462033176  a18e0999-61a1-4218-8f8f-61ec1d375361   
2019-10-01 00:01:55  514753614  e2fecb2d-22d0-df2c-c661-15da44b3ccf1   
2019-10-01 00:02:50  527418424  86e77869-afbc-4dff-9aa2-6b7dd8c90770   
2019-10-01 00:03:41  555448072  b5f72ceb-0730-44de-a932-d16db62390df   
2019-10-01 00:03:44  552006247  2d8f304b-de45-4e59-8f40-50c603843fe5   
...                        ...                                   ...   
2020-02-29 23:58:49  147995998  5ff96629-3627-493e-a25b-5a871ec78c90   
2020-02-29 23:58:57  147995998  5ff96629-3627-493e-a25b-5a871ec78c90   
2020-02-29 23:59:05  147995998  5ff96629-3627-493e-a25b-5a871ec78c90   
2020-02-29 23:59:28  619841242  18af673b-7fb9-4202-a66d-5c855bc0fd2d   
2020-02-29 23:59:54  619841242  18af673b-7fb9-4202-a66d-5c855bc0fd2d   

                             category_id event_type  product_id  \
event_time                                                        
2019-10-01 00:01:46  1487580005092295511       view     5843665   
2019-10-01 00:01:55  1487580013069861041       cart     5868461   
2019-10-01 00:02:50  1487580006300255120       view     5877456   
2019-10-01 00:03:41  1487580013749338323       view     5649270   
2019-10-01 00:03:44  1487580005411062629       view       18082   
...                                  ...        ...         ...   
2020-02-29 23:58:49  1487580006317032337       cart     5815662   
2020-02-29 23:58:57  1487580006317032337       view     5815665   
2020-02-29 23:59:05  1487580006317032337       cart     5815665   
2020-02-29 23:59:28  1487580010872045658       view     5817692   
2020-02-29 23:59:54  1487580010872045658       view     5716351   

                              price        date  year  month  day  hour  \
event_time                                                                
2019-10-01 00:01:46            9.44  2019-10-01  2019     10    1     0   
2019-10-01 00:01:55            3.57  2019-10-01  2019     10    1     0   
2019-10-01 00:02:50          122.22  2019-10-01  2019     10    1     0   
2019-10-01 00:03:41            6.19  2019-10-01  2019     10    1     0   
2019-10-01 00:03:44           16.03  2019-10-01  2019     10    1     0   
...                             ...         ...   ...    ...  ...   ...   
2020-02-29 23:58:49            0.92  2020-02-29  2020      2   29    23   
2020-02-29 23:58:57            0.59  2020-02-29  2020      2   29    23   
2020-02-29 23:59:05            0.59  2020-02-29  2020      2   29    23   
2020-02-29 23:59:28            0.79  2020-02-29  2020      2   29    23   
2020-02-29 23:59:54            0.79  2020-02-29  2020      2   29    23   

                     minute  second  holidays  black_friday  valentine_day  
event_time                                                                  
2019-10-01 00:01:46       1      46         0             0              0  
2019-10-01 00:01:55       1      55         0             0              0  
2019-10-01 00:02:50       2      50         0             0              0  
2019-10-01 00:03:41       3      41         0             0              0  
2019-10-01 00:03:44       3      44         0             0              0  
...                     ...     ...       ...           ...            ...  
2020-02-29 23:58:49      58      49         0             0              0  
2020-02-29 23:58:57      58      57         0             0              0  
2020-02-29 23:59:05      59       5         0             0              0  
2020-02-29 23:59:28      59      28         0             0              0  
2020-02-29 23:59:54      59      54         0             0              0  

[2074026 rows x 16 columns]

In [56]:
df.to_pickle('..\Datos\Analytic_datamart.pickle')